### PR TITLE
Add KVM_SET_TSC_KHZ and KVM_GET_TSC_KHZ wrappers.

### PR DESCRIPTION
**Pull request by sandreim**
Add wrappers needed to enable TSC scaling for CPUs that support it.

Implemented two vCPU wrappers:
- get_tsc_khz() for KVM_GET_TSC_KHZ - depends on KVM cap KVM_CAP_GET_TSC_KHZ
- set_tsc_khz() for KVM_SET_TSC_KHZ - depends on KVM cap KVM_CAP_TSC_CONTROL

### CODE REVIEW

File: src/ioctls/vcpu.rs [hunk 1]
```diff
@@ -2172,4 +2205,35 @@ mod tests {
             vcpu.enable_cap(&cap).unwrap();
         }
     }
+    #[cfg(target_arch = "x86_64")]
+    #[test]
+    fn test_get_tsc_khz() {
```
Comment (andreeaflorescu) [inline src/ioctls/vcpu.rs:2234]: What is the reason for having this as a separate test? I was thinking we can test both the setter and getter in the same test (something like `test_tsc_khz`) because otherwise this test seems trivial, and the functionality would be covered in  `test_set_tsc_khz`.

File: src/kvm_ioctls.rs [hunk 2]
```diff
@@ -1,3 +1,4 @@
+//! A safe wrapper around the kernel's KVM interface.
```
Comment (andreeaflorescu) [inline src/kvm_ioctls.rs]: This change does not seem to be related to the PR.

File: src/ioctls/vcpu.rs [hunk 3]
```diff
@@ -1339,6 +1339,39 @@ impl VcpuFd {
...
+    ///
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    pub fn get_tsc_khz(&self) -> Result<u32> {
+        // Safe because we know that our file is a KVM fd and that the request is one of the ones
+        // defined by kernel.
+        let ret = unsafe { ioctl(self, KVM_GET_TSC_KHZ()) };
+        if ret >= 0 {
+            Ok(ret as u32)
+        } else {
+            Err(errno::Error::new(ret))
```
Comment (andreeaflorescu) [inline src/ioctls/vcpu.rs:1364]: Can you also add a negative test?

File: src/ioctls/vcpu.rs [hunk 4]
```diff
@@ -1339,6 +1339,39 @@ impl VcpuFd {
...
+        }
+    }
+
+    /// Sets the specified vCPU TSC frequency.
```
Comment (andreeaflorescu) [inline src/ioctls/vcpu.rs:1368]: We are trying to add usage examples as well for the public interface where this is not too hard to achieve.

**Review by andreeaflorescu — COMMENTED** [review]
Just curious, why do we need these two ioctls? What functionality are they helping with? (maybe we can write something about it in the commit message as well).

File: src/ioctls/vcpu.rs [hunk 5]
```diff
@@ -2172,4 +2205,35 @@ mod tests {
...
+        let vcpu = vm.create_vcpu(0).unwrap();
+        assert!(vcpu.get_tsc_khz().unwrap() > 0);
+    }
+
+    #[cfg(target_arch = "x86_64")]
+    #[test]
+    fn test_set_tsc_khz() {
+        let kvm = Kvm::new().unwrap();
+        // Skip test if KVM/host capabilities are missing.
+        if !(kvm.check_extension(Cap::GetTscKhz) && kvm.check_extension(Cap::TscControl)) {
```
Comment (lauralt) [inline src/ioctls/vcpu.rs]: For the hosts that are missing these capabilities we could test that `set/get` return an error.

File: src/ioctls/vcpu.rs (hunk 1 above)
Comment (lauralt) [reply src/ioctls/vcpu.rs:2234]: I think it's fine here to have separate tests since on some machines (like my laptop for example) `GetTscKhz` is present and `TscControl` is missing.
We should anyway test in `test_set_tsc_khz` that `get_tsc_khz` returns the value that was just set.

File: src/ioctls/vcpu.rs (hunk 3 above)
Comment (georgepisaltu) [reply src/ioctls/vcpu.rs:1364]: Unfortunately, there is no way to reproduce the negative outcome of the function in a reliable way in the test. I think we would need mock structs for this purpose.

I updated the tests to check for machines where these capabilities are not present though.

File: src/ioctls/vcpu.rs (hunk 4 above)
Comment (georgepisaltu) [reply src/ioctls/vcpu.rs:1368]: Done.

File: src/ioctls/vcpu.rs (hunk 1 above)
Comment (georgepisaltu) [reply src/ioctls/vcpu.rs:2234]: Done.

File: src/ioctls/vcpu.rs (hunk 5 above)
Comment (georgepisaltu) [reply src/ioctls/vcpu.rs]: Done.

File: src/kvm_ioctls.rs (hunk 2 above)
Comment (georgepisaltu) [reply src/kvm_ioctls.rs]: Removed.

File: src/ioctls/vcpu.rs (hunk 3 above)
Comment (andreeaflorescu) [reply src/ioctls/vcpu.rs:1364]: Can we test this with a faulty vcpu fd? Here is the test: https://github.com/rust-vmm/kvm-ioctls/blob/9d0fd1fbc22918fdd03c89ca888195a8ad297685/src/ioctls/vcpu.rs#L1910

Comment (georgepisaltu) [reply src/ioctls/vcpu.rs:1364]: I tried using that example, PTAL.

File: src/ioctls/vcpu.rs [hunk 6]
```diff
@@ -2172,4 +2227,57 @@ mod tests {
...
+        use std::os::unix::io::FromRawFd;
+
+        let faulty_vcpu_fd = VcpuFd {
+            vcpu: unsafe { File::from_raw_fd(-1) },
+            kvm_run_ptr: KvmRunWrapper {
+                kvm_run_ptr: mmap_anonymous(10),
+                mmap_size: 10,
+            },
+        };
+        assert!(faulty_vcpu_fd.get_tsc_khz().is_err());
```
Comment (andreeaflorescu) [inline src/ioctls/vcpu.rs]: Can you just add this call to the already existing test function so that we do not need to create another faulty_vcpu_fd? Sorry, I wasn't being explicit in my previous comment.

Comment (georgepisaltu) [reply src/ioctls/vcpu.rs]: Yes, I moved it to the other test.

**Review by lauralt — APPROVED** [review]

**Review by andreeaflorescu — APPROVED** [review]
